Update orders swagger spec and fix api_key access to orders endpoint

mbudm opened the pull request:
#### What? Why?

Closes #3770

This updates the swagger specification for the OFN API orders endpoint. It documents the query parameters that already exist on this endpoint. It also includes a change from this draft PR that enables access to the /api/orders endpoint via an api key.

I wanted to also add tests for the API, however I've run into some difficulty covered in this issue comment. This PR can be reviewed and we can decide to accept it or wait on tests being created

#### What should we test?
Test that the `/api/orders` endpoint works via the browser with a logged in user **and** via POSTMAN or curl with an api key - [api authorization guide](https://github.com/openfoodfoundation/openfoodnetwork/wiki/API-documentation#authorization).
Test that the query parameters that are documented for `/api/orders` work as described. The query parameters are:
```
q[completed_at_gt]:2020-05-27
q[completed_at_lt]:2020-05-28
q[distributor_id_eq]:4
q[state_eq]:complete
q[payment_state_eq]:balance_due
q[order_cycle_id_eq]:2
q[ready_to_capture_payment_payment_method_id_eq]:1
q[email_cont]:@gmail
```

#### Release notes
Fixed access to the api/orders endpoint with API key
Documented the api/orders query parameters

Changelog Category: Fixed

mbudm commented:
I've applied the approach to using rswag tests sugested by @luisramos0  - currently over in this [WIP commit](https://github.com/mbudm/openfoodnetwork/commit/a69fe1f2a8176448e60490f1ce2ce3e8fed41373)

I'll roll that into here once I resolve why those tests are failing. I suggest moving [that discussion to slack](https://openfoodnetwork.slack.com/archives/C2GQ45KNU/p1592358390302700?thread_ts=1591757249.297100&cid=C2GQ45KNU) as it's more about my lack of rspec / api knowledge.

Once those tests are in this PR we will:
- have tests for the orders endpoint, meaning we can include the code change originally in #5475
- have tests for a selection of query parameters, that we are comfortable supporting right now (we can add more later, as needed)
- **not have** autogenerated swagger docs yet
  - as there are a couple of changes to rswag needed to enable valid swagger 3.0.1 docs (see comments in WIP commit)
  - because all endpoints need test coverage before we can auto generate a complete swagger spec
- have manually updated swagger doc covering the query parameters that are now tested

So the roadmap to autogenerated swagger docs is:
- add sample rswag test (this PR, including WIP commit)
- add issue to check rswag changes when ready to check valid 3.0.1 is generated
- add tests for other api endpoints
- when autogenerated spec matches the existing manual spec, delete the manual version and replace with the autogenerated version
- document autogeneration process (or perhaps automate it)

luisramos0 commented on spec/requests/api/orders_spec.rb:
```diff
@@ -0,0 +1,130 @@
+          let(:order_dist_2) { create(:order_with_distributor) }
+          let(:order_dist_1_complete) { create(:order,  distributor: order_dist_1.distributor, state: 'complete', completed_at: Time.zone.today - 7.days) }
+          let(:order_dist_1_credit_owed) { create(:order,  distributor: order_dist_1.distributor, state: 'complete', payment_state: 'credit_owed', completed_at: Time.zone.today) }
+          
```
I gave this a try, you need to use let! to get the orders created.

You can debug SearchOrders#fetch_orders for details.
This (using let!) gets 3 orders into the first test:
![image](https://user-images.githubusercontent.com/1640378/84902719-16648400-b0a5-11ea-9112-797043b74d76.png)


The 4th order is not showing up because the user has no permissions to edit  order_dist_2. If you want to get the 4 orders in the first test you need to make the user a manager or owner of order_dist_2.distributor

luisramos0 replied:
with this extra setup the first test passes:
before { order_dist_2.distributor.update_attributes owner: user }

mbudm replied:
Thanks - I knew it was gonna be something stupidly obvious. In my travels I did see a section on difference between let and let! but didn't read it

luisramos0 replied:
:+1:

luisramos0 replied:
Did I say I had zero experience with ruby and rails when I started in OFN 2 years ago? I mean, I have been there!
Code Review is very valuable, my advise is: dont spend too much time when you are stuck, get it out in a PR and someone with more rails on their hands will fix it for you. This was a good example :+1: 
The debugging part is important, get the byebug instruction in the right place and when there, see what's going on, in this case, I run editable_orders in SearchOrders#fetch_orders and then, because the orders where not showing there I run Spree::Orders.all and saw the orders were not created.

mbudm replied:
Really, I'd never have guessed you were a ruby newbie so recently @luisramos0  - that's good to know!

I was thinking also I should have just checked the test DB whilst debugging the test - something I'm much more familiar with.

I'm definitely happy to share early and often. I am way past caring about looking stupid after this many years in dev ;-)

luisramos0 commented:
I think you need to rebase to get latest master and see if this still works in rails 4 :-)

mbudm commented:
Updates today, I think this resolves all review comments:
- Tests fixed thanks to @luisramos0 great tips.
- I have committed the auto generated swagger spec (and added a note to the description to avoid this being used prematurely)
- I've removed the rswag UI and API engines as we don't want to host the swagger spec with the app, just yet.

I think the next steps, after merge, are):
- update swagger spec at https://app.swaggerhub.com/apis/luisramos0/the-open_food_network/0.1 (or change to a generic account
- document this use of rswag 
- add issue to OFN noting the updates of rswag that are pending and needed to support open api 3.0.1
- add issue to OFN to remove static swagger when api test coverage is complete enough that we can auto generate a swagger spec from the tests.

luisramos0 commented on swagger/v1/swagger.yaml:
```diff
@@ -0,0 +1,124 @@
+info:
+  title: The Open Food Network
+  description: |-
+    This spec is auto generated using the rswag gem. It is incomplete and not yet valid for openapi 3.0.1. Do not publish this. \
```
"Do not publish this."
It's ok if you think this is the best thing to do now.
I feel we can just move forward as much as we can already. I'd just keep the engine and have this public. We can then improve it.

mbudm replied:
For now our API documentation 'process' is to update the swagger.yaml in root and then update https://app.swaggerhub.com/. I think we should only add things to the repo that are in use, not that may be used later, no matter how certain.

For me that means not adding the engine, as it's not in use yet. 

Rswag on the other hand is being used as the API spec is active, and I don't mind the half baked new swagger.yaml being committed either, as we'll now be able to see progress on it as more tests are added, and rswag issues are fixed. So arguably that is in use -- it serves a purpose.

luisramos0 replied:
ah, ok, let's see if we are on the same page here.

"For now our API documentation 'process' is to update the swagger.yaml in root and then update "

I thought from now on the api doc process is to write a spec and rerun rswag? Is it not?
I thought the engine would publish the latest version of that doc, which is incomplete. The manual swagger doc is also incomplete.

luisramos0 replied:
I think I am saying I'd be happy to deprecate the manual process already. To add context, it has been used at all, it was created by a group of students and updated by me once.

luisramos0 replied:
I don't have a strong opinion about this! I am just trying to clarify where we are going with this :+1:

luisramos0 replied:
Actually, never mind, from reading your todo list above I understand what's your plan :+1:

mbudm commented on swagger.yaml:
```diff
@@ -342,9 +342,63 @@ paths:
+          description: Query orders completed after a date. 
+          required: false
+        - in: query
+          name: q[completed_at_lt]
```
We've got to be brief and accurate in order to make the spec readable, so I think the description captures it:

`Query orders completed before a date.`

If that changes, we should change the description.

luisramos0 commented:
ok, I run the tests, all looks good except the filter ready_to_capture_payment_payment_method_id_eq, it doesnt work, I dont see any specific change related to this filter in the PR. So I will consider this Pr done and I will merge.

I investigated the ready_to_capture_payment_payment_method_id_eq filter and I found that the property order.ready_to_capture should be a boolean but it is actually serializing a payment! I think we need to change this in the serializer:
![image](https://user-images.githubusercontent.com/1640378/86632682-9d7f8a80-bfc7-11ea-8b83-ddb7d64ccd7b.png)

@mbudm I wonder why/how this ready_to_capture_payment_payment_method_id_eq filter was considered for testing.
